### PR TITLE
Shape inference for Theano backend

### PR DESCRIPTION
**Pull request by abhaikollara**
Shape inferences added

| Op |Test coverage|
|-----------------|:-----------------:|
| `flatten` | :heavy_check_mark: |
| `repeat_elements` | :heavy_check_mark: |
| `batch_flatten` | :heavy_check_mark: |
| `squeeze` | :heavy_check_mark: |
| `permute_dimensions` | :heavy_check_mark: |
| `repeat` | :heavy_check_mark: |
| `expand_dims` | :heavy_check_mark: |
| `spatial_2d_pooling` |
|`transpose`|:heavy_check_mark: |
|`batch_dot`|:heavy_check_mark: |

The original PR was at https://github.com/fchollet/keras/pull/5610. The suggested changes have been fixed. Had to remove shape inference for `tile` due to certain issues (still working on it)

### CODE REVIEW

**Comment by fchollet** [comment]
> The suggested changes have been fixed

What about unit tests (as part of backend tests, theano-only)?

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -359,12 +358,26 @@ def batch_dot(x, y, axes=None):
     out = T.batched_tensordot(x, y, axes=axes)
     if ndim(out) == 1:
         out = expand_dims(out, 1)
+
+    if hasattr(x, '_keras_shape') and hasattr(y, '_keras_shape'):
+        shape = []
+        for axis in range(len(x._keras_shape)):
+            if axis != axes[0]:
+                shape.append(x._keras_shape[axis])
+        for axis in range(1,len(y._keras_shape)):
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: PEP8 issue. You should use a PEP8 linter, it would make your life easier. https://pypi.python.org/pypi/pep8

**Comment by abhaikollara** [comment]
I've added some tests for `keras_shape` inference. It should cover `reshape`,`permute_dimensions`,`repeat`,`flatten`,`expand_dims`,`squeeze`,`batch_dot`,`dot`,`transpose`

File: keras/backend/theano_backend.py (hunk 1 above)
Comment (abhaikollara) [inline keras/backend/theano_backend.py]: Thanks :smile: I'll take that advice

**Comment by fchollet** [comment]
LGTM. Thanks a lot!

**Comment by around1991** [comment]
Hi, sorry to be a pain, but a lot of the current shape inference breaks if there's a None shape dimension in the input.

For example, if `x` is a variable size 2D tensor with shape (None, None), then `K.tile(x, (5, 1))` should also infer shape None. I know the Keras philosophy seems to be to specify as much information as possible during graph compilation time, but sometimes it's useful to be able to do shape inference at run time with the underlying backend, rather than forcing us to specify exact shapes of batches, for instance.

**Comment by farizrahman4u** [comment]
Right. Too many bugs.

**Pull request merged**